File: validateBinarySearchTree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def __init__(self):
        self.pre = None
        
    def isValidBST(self, root):
        return self.isBST(root)
        
    def isBST(self,node):
        if not node:
            return True
        if not self.isBST(node.left):
            return False
        if self.pre and self.pre.val >= node.val:
            return False
        self.pre = node
        return self.isBST(node.right)

File: JAVA/validateBinarySearchTree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private TreeNode pre = null;
    
    public boolean isValidBST(TreeNode root) {
        return inorderCheck(root);
    }
    public boolean inorderCheck(TreeNode root){
        if (root == null)
            return true;
        if (!inorderCheck(root.left))
            return false;
        if (pre != null && pre.val >= root.val)
            return false;
        pre = root;
        return inorderCheck(root.right);
    }
}